Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -385,38 +428,46 @@ private Block NextBlockCore(CancellationToken cancellationToken)
                 {
                     if (header.Previous.HashBlock != this.location.HashBlock)
                     {
-                        //A reorg
-                        return null;
+                        this.logger.LogTrace("Blockchain reorganization detected.");
+                        break;
                     }
                     this.IsStalling = false;
                     this.location = header;
                     Interlocked.Add(ref this.currentSize, -block.Length);
                     this.consumed.Set();
-                    return block.Block;
+
+                    res = block.Block;
                 }
                 else
                 {
-                    // Otherwise we either have reorg.
+                    // Otherwise we either have reorg, or we don't know the next block.
                     if (header == null)
                     {
                         if (!this.Chain.Contains(this.location.HashBlock))
                         {
-                            //A reorg
-                            return null;
+                            this.logger.LogTrace("Blockchain reorganization detected.");
+                            break;
                         }
+
+                        this.logger.LogTrace("Hash of the next block is not known.");
                     }
                     else
                     {
+                        this.logger.LogTrace($"Block not available.");
+
                         // Or the block is still being downloaded or we need to ask for this block to be downloaded.
                         if (!isDownloading) AskBlocks(new ChainedBlock[] { header });
 
                         OnStalling(header);
                         this.IsStalling = true;
                     }
-                    WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, waitTime[i]);
-                    i = i == waitTime.Length - 1 ? 0 : i + 1;
+
+                    WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, WaitNextBlockRoundTimeMs);
```

Comment:
I understand why you are asking for this - to avoid useless work ("are we there yet polling") to be done every so often, but as I look in the code, those checks are all O(1), i.e. we are talking about sub millisecond work again, which should be perfectly fine doing every 100ms, especially when the chain is synced already (i.e. not much work is actually in progress elsewhere)

this is why I would stay with the current simpler design - i don't think the change is worth it